cupyx/scipy/signal/_peak_finding.py

```diff
@@ -1492,3 +1497,659 @@ def argrelextrema(data, comparator, axis=0, order=1, mode="clip"):
+    ...                  [  43., -999., -999.],
+    ...                  [  40., -999., -999.],
+    ...                  [  36., -999., -999.]])
+    >>> print(mquantiles(data, axis=0, limit=(0, 50)))
+    [[19.2  14.6   1.45]
+     [40.   37.5   2.5 ]
+     [42.8  40.05  3.55]]
+
+    >>> data[:, 2] = -999.
+    >>> print(mquantiles(data, axis=0, limit=(0, 50)))
+    [[19.200000000000003 14.6 --]
+     [40.0 37.5 --]
+     [42.800000000000004 40.05 --]]
+
+    """
+    def _quantiles1D(data, m, p):
+        x = np.sort(data.compressed())
+        n = len(x)
+        if n == 0:
+            return ma.array(np.empty(len(p), dtype=float), mask=True)
+        elif n == 1:
+            return ma.array(np.resize(x, p.shape), mask=nomask)
+        aleph = (n*p + m)
+        k = np.floor(aleph.clip(1, n-1)).astype(int)
+        gamma = (aleph-k).clip(0, 1)
+        return (1.-gamma)*x[(k-1).tolist()] + gamma*x[k.tolist()]
+
+    data = ma.array(a, copy=False)
+    if data.ndim > 2:
+        raise TypeError("Array should be 2D at most !")
+
+    if limit:
+        condition = (limit[0] < data) & (data < limit[1])
+        data[~condition.filled(True)] = masked
+
+    p = np.array(prob, copy=False, ndmin=1)
+    m = alphap + p*(1.-alphap-betap)
+    # Computes quantiles along axis (or globally)
+    if (axis is None):
+        return _quantiles1D(data, m, p)
+
+    return ma.apply_along_axis(_quantiles1D, axis, data, m, p)
+
+
+def _scoreatpercentile(data, per, limit=(), alphap=.4, betap=.4):
+    """Calculate the score at the given 'per' percentile of the
+    sequence a.  For example, the score at per=50 is the median.
+
+    This function is a shortcut to mquantile
+
+    """
+    if (per < 0) or (per > 100.):
+        raise ValueError("The percentile should be between 0. and 100. !"
+                         " (got %s)" % per)
+
+    return _mquantiles(data, prob=[per/100.], alphap=alphap, betap=betap,
+                       limit=limit, axis=0).squeeze()
+
+
+_NOISE_MODULE_SRC = r"""
+#include <cstddef>  // for std::size_t
+#include <type_traits>
+
+#include <thrust/sort.h>
+#include <thrust/functional.h>
+#include <thrust/execution_policy.h>
+
+__device__ inline float clip(float value, float lb, float ub) {
+    return value < lb ? lb : ((value > ub) ? ub : value);
+}
+
+template<typename T>
+__global__ cwt_noise(
+        const int n_points,
+        const int window_size,
+        const float prob,
+        const float m,
+        const T* __restrict__ cwt_row,
+        T* out) {
+
+    int idx = blockDim.x * blockIdx.x + threadIdx.x;
+
+    if(idx >= n_points) {
+        return;
+    }
+
+    const int win_half_size = window_size / 2;
+    const int odd = window_size % 2;
+
+    const int left_pos = min(idx - win_half_size, 0);
+    const int right_pos = min(idx + win_half_size + odd, n);
+    const int n = right_pos - left_pos;
+
+    thrust::sort(thrust::device, cwt_row + left_pos, cwt_row + right_pos);
```

Comment:
I don't understand what this line is doing? I get that you want each thread to sort the data within a window, and that the window is sliding across the thread ID. However,  is doing in-place sort, meaning each thread is writing data back to the same buffer, and the buffers seen by adjacent threads have substantial overlapping, thus leading to a potential race condition. Am I missing something?

Reply:
Also, for issues with using `thrust::device` in a kernel, see https://github.com/cupy/cupy/issues/8239.

I would suggest to rewrite this kernel to use CUB block-level sort (merge sort or radix sort), and avoid using Thrust.

Reply:
Alternatively, to retain the same kernel as-is we could implement a homegrown serial sort here (maybe it's already somewhere in the codebase?), but it's unclear to me what the perf impact would be.

Reply:
xref: https://github.com/NVIDIA/cccl/blob/main/cub/cub/thread/thread_sort.cuh

Reply:
Thanks for your timely review @leofang, I'll come back to this PR and try to do a per-thread sort and see if it improves the performance.

Reply:
Note that the scope of thrust sort here is also per-thread, so there's no much different.